adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java

```diff
@@ -645,34 +666,91 @@ protected void preShutdown() {
                     .withTag(TracingHelper.TAG_AUTHENTICATED.getKey(), authenticatedDevice != null)
                     .start();
 
+            final Promise<Void> responseReady = Promise.promise();
+
             final Future<JsonObject> tokenTracker = getRegistrationAssertion(
                     device.getTenantId(), device.getDeviceId(),
                     authenticatedDevice,
                     currentSpan.context());
-            final Future<TenantObject> tenantTracker = getTenantConfiguration(device.getTenantId(), currentSpan.context());
+            final Future<TenantObject> tenantTracker = getTenantConfiguration(device.getTenantId(),
+                    currentSpan.context());
             final Future<TenantObject> tenantValidationTracker = tenantTracker
                     .compose(tenantObject -> CompositeFuture
                             .all(isAdapterEnabled(tenantObject),
                                     checkMessageLimit(tenantObject, payload.length(), currentSpan.context()))
                             .map(success -> tenantObject));
+
+            // we only need to consider TTD if the device and tenant are enabled and the adapter
+            // is enabled for the tenant
+            final Future<Integer> ttdTracker = CompositeFuture.all(tenantValidationTracker, tokenTracker)
+                    .compose(ok -> {
+                        final Integer ttdParam = getTimeUntilDisconnectFromRequest(context);
+                        return getTimeUntilDisconnect(tenantTracker.result(), ttdParam).map(effectiveTtd -> {
+                            if (effectiveTtd != null) {
+                                currentSpan.setTag(MessageHelper.APP_PROPERTY_DEVICE_TTD, effectiveTtd);
+                            }
+                            return effectiveTtd;
+                        });
+                    });
+            final Future<MessageConsumer> commandConsumerTracker = ttdTracker
```

Comment:
It looks like the `commandConsumerTracker` is not used below in the sense that it is `compose`d with other Futures, to handle command consumer creation failures.

If I understand the intention here correctly, the command and control handling in `doUploadMessage` is similar to the one in the HTTP adapter, meaning that when the returned Future of  `doUploadMessage`  is finished, the created command consumer is to be closed again already, right?
The line below:
`Optional.ofNullable(commandConsumerTracker.result()).ifPresent(consumer -> consumer.close(null));`
means: If the command consumer creation was successfully finished, close the consumer.
But what if at that point in time, the consumer creation wasn't finished **yet**. When the creation succeeds at a later point, the consumer will never get closed.

Therefore the code here has to be adapted to make sure that a successfully created command consumer is closed under all circumstances. That means, making sure that some logic is done *after* the `commandConsumerTracker` is completed (using `commandConsumerTracker.compose` or `CompositeFuture.join` as done [here](https://github.com/eclipse/hono/blob/bbacfb9d8797e9e3ba95dcc75166ce1b7647fd23/adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java#L697)).

Reply:
Hopefully fixed :-).